iohub/dataset_base.py

```diff
@@ -0,0 +1,71 @@
+from abc import ABC, abstractmethod
+from typing import Generator, Union
+
+import numpy as np
+from numpy.typing import ArrayLike
+
+
+class BaseFOV(ABC):
+    # NOTE: not using the `data` method from ngff Position
+
+    @property
+    @abstractmethod
+    def channels(self) -> list[str]:
+        raise NotImplementedError
+
+    def channel_index(self, key: str) -> int:
+        """Return index of given channel."""
+        return self.channels.index(key)
+
+    @property
+    @abstractmethod
+    def shape(self) -> tuple[int, int, int, int, int]:
+        # NOTE: suggestion, fix dimension to 5?
+        # We could me more restrictive than ome-zarr
+        raise NotImplementedError
+
+    @abstractmethod
+    def __getitem__(
+        self,
+        key: Union[int, slice, tuple[Union[int, slice], ...]],
+    ) -> ArrayLike:
+        """
+        Output object must support __array__ interface, np.asarray(...).
+        """
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def ndim(self) -> int:
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def dtype(self) -> np.dtype:
+        raise NotImplementedError
+
+    @property
+    def scale(self) -> list[float]:
+        """Helper function for FOV scale."""
+        raise NotImplementedError
+
+
+class BaseFOVCollection(ABC):
+    @abstractmethod
+    def __contains__(self, position_key: str) -> bool:
+        """Check if a position is present in the collection."""
+        raise NotImplementedError
+
+    @abstractmethod
+    def __len__(self) -> int:
+        raise NotImplementedError
+
+    @abstractmethod
+    def __getitem__(self, position_key: str) -> BaseFOV:
+        """FOV key position to FOV object."""
+        raise NotImplementedError
+
+    @abstractmethod
+    def __iter__(self) -> Generator[tuple[str, BaseFOV], None, None]:
+        # NOTE is this preferred than the current `.positions` ?
```

Comment:
`NGFFNode` also has `__iter__`, but that only iterates on the immediate next level in the hierarchy (e.g. rows in plate, wells in row etc). The `rows()`, `wells()`, and `positions()` methods allows for iteration on different levels from the same starting point (e.g. plate).

Reply:
I think it is still a good idea to always yield `BaseFOV` here. So we will need to change the behavior of the NGFF plate node for compliance.

Reply:
Ok, can I leave this to be implemented in a different PR when we inherit the base classes on the `NGFFNode`?

Reply:
Sounds good! Actually I should take care of it once this merges.